src/Common/src/System/Net/Internals/IPEndPointExtensions.cs

```diff
@@ -9,28 +9,63 @@ internal static class IPEndPointExtensions
     {
         public static Internals.SocketAddress Serialize(EndPoint endpoint)
         {
-            Debug.Assert(endpoint is IPEndPoint);
+            var ipEndPoint = endpoint as IPEndPoint;
+            if (ipEndPoint != null)
+            {
+                return new Internals.SocketAddress(ipEndPoint.Address, ipEndPoint.Port);
+            }
 
-            return new Internals.SocketAddress(((IPEndPoint)endpoint).Address, ((IPEndPoint)endpoint).Port);
+            System.Net.SocketAddress address = endpoint.Serialize();
```

Comment:
There are already existing tests for DnsEndpoint. The code checks if it's DNS end point before it calls IPEndPointExtensions.